File: AccuPay/Core Forms/Commons/ListOfValueCollection.vb
Option Strict On

Imports AccuPay.Entity
Imports System.Collections.ReadOnlyCollectionBase
Imports System.Collections.Generic
Imports System.Collections.ObjectModel

Public Class ListOfValueCollection

    Private _values As IReadOnlyList(Of ListOfValue)

    Public Sub New(values As IEnumerable(Of ListOfValue))
        _values = New ReadOnlyCollection(Of ListOfValue)(values.ToList)
    End Sub

    Public Function Exists(lic As String) As Boolean
        Dim value = _values?.FirstOrDefault(Function(f) f.LIC = lic)
        Return value IsNot Nothing
    End Function

    Public Function GetValue(lic As String) As String
        Dim value = _values?.FirstOrDefault(Function(f) f.LIC = lic)
        Return value?.DisplayValue
    End Function

    Public Function GetValue(type As String, lic As String) As String
        Dim value = _values?.FirstOrDefault(Function(f) f.Type = type And f.LIC = lic)
        Return value?.DisplayValue
    End Function

    Public Function GetSublist(type As String) As ListOfValueCollection
        Return New ListOfValueCollection(_values.Where(Function(l) l.Type = type))
    End Function

    Public Function GetString(name As String, Optional [default] As String = "") As String
        Dim names = Split(name)
        Return GetStringValue(names.Item1, names.Item2)
    End Function

    Public Function GetStringOrNull(name As String) As String
        Dim names = Split(name)

        Return GetStringValue(names.Item1, names.Item2)
    End Function

    Public Function GetStringOrDefault(name As String, Optional [default] As String = "") As String
        Dim names = Split(name)

        Return If(GetStringValue(names.Item1, names.Item2), [default])
    End Function

    Private Function GetStringValue(type As String, lic As String) As String
        Dim value = _values?.FirstOrDefault(Function(f) f.LIC = lic And f.Type = type)
        Return value?.DisplayValue
    End Function

    Public Function GetEnum(Of T As {Structure})(name As String, Optional [default] As T = Nothing) As T
        Dim names = Split(name)
        Return GetEnum(Of T)(names.Item1, names.Item2, [default])
    End Function

    Public Function GetEnum(Of T As {Structure})(type As String, lic As String, Optional [default] As T = Nothing) As T
        Dim value = GetValue(type, lic)

        If value Is Nothing Then
            Return [default]
        End If

        Return Enums(Of T).Parse(value)
    End Function

    Public Function GetBoolean(name As String, Optional [default] As Boolean = False) As Boolean
        Dim names = Split(name)
        Return GetBoolean(names.Item1, names.Item2, [default])
    End Function

    Public Function GetBoolean(type As String, lic As String, Optional [default] As Boolean = False) As Boolean
        Dim value = GetListOfValue(type, lic)

        If String.IsNullOrEmpty(value?.DisplayValue) Then
            Return [default]
        End If

        If IsNumeric(value?.DisplayValue) Then
            Return Convert.ToBoolean(
                Convert.ToInt32(value?.DisplayValue))
        End If

        Return Convert.ToBoolean(value?.DisplayValue)
    End Function

    Public Function GetDecimal(lic As String, Optional [default] As Decimal = 0) As Decimal
        Return If(GetDecimalOrNull(lic), [default])
    End Function

    Public Function GetDecimalOrNull(lic As String) As Decimal?
        Dim value = GetValue(lic)
        Return If(value IsNot Nothing, Decimal.Parse(value), Nothing)
    End Function

    Private Function Split(name As String) As Tuple(Of String, String)
        Dim names = name.Split({"."}, 2, StringSplitOptions.RemoveEmptyEntries)
        Dim type = names(0)
        Dim lic = names(1)

        Return New Tuple(Of String, String)(type, lic)
    End Function

    Private Function GetListOfValue(type As String, lic As String) As ListOfValue
        Return _values?.FirstOrDefault(Function(f) f.Type = type And f.LIC = lic)
    End Function

End Class

File: Accupay.Data/Services/ListOfValueCollection.cs
using AccuPay.Data.Entities;
using AccuPay.Data.Helpers;
using AccuPay.Utilities;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;

namespace AccuPay.Data.Services
{
    public class ListOfValueCollection
    {
        private IReadOnlyList<ListOfValue> _values;

        public ListOfValueCollection(IEnumerable<ListOfValue> values)
        {
            _values = new ReadOnlyCollection<ListOfValue>(values.ToList());
        }

        public bool Exists(string lic)
        {
            var value = _values?.FirstOrDefault(f => f.LIC == lic);
            return value != null;
        }

        public string GetValue(string lic, bool findByOrganization = false, int? organizationId = null)
        {
            ListOfValue value;

            if (findByOrganization)
            {
                value = _values?.FirstOrDefault(f => f.LIC == lic &&
                                                f.OrganizationID != null &&
                                                f.OrganizationID.Value == organizationId);

                if (value == null)
                    value = _values?.FirstOrDefault(f => f.LIC == lic);
            }
            else
                value = _values?.FirstOrDefault(f => f.LIC == lic);

            return value?.DisplayValue;
        }

        public string GetValue(string type, string lic, bool findByOrganization = false, int? organizationId = null)
        {
            ListOfValue value;

            if (findByOrganization)
            {
                value = _values?.FirstOrDefault(f => f.Type == type &&
                                                    f.LIC == lic &&
                                                    f.OrganizationID != null &&
                                                    f.OrganizationID.Value == organizationId);

                if (value == null)
                    value = _values?.FirstOrDefault(f => f.Type == type && f.LIC == lic);
            }
            else
                value = _values?.FirstOrDefault(f => f.Type == type && f.LIC == lic);

            return value?.DisplayValue;
        }

        public ListOfValueCollection GetSublist(string type)
        {
            return new ListOfValueCollection(_values.Where(l => l.Type == type));
        }

        public string GetString(string name, string @default = "")
        {
            var names = Split(name);
            var value = GetStringValue(names.Item1, names.Item2);

            return value ?? @default;
        }

        public string GetStringOrNull(string name)
        {
            var names = Split(name);

            return GetStringValue(names.Item1, names.Item2);
        }

        public string GetStringOrDefault(string name, string @default = "")
        {
            var names = Split(name);

            return GetStringValue(names.Item1, names.Item2) ?? @default;
        }

        public T GetEnum<T>(string name, T @default = default(T), bool findByOrganization = false) where T : struct
        {
            var names = Split(name);
            return GetEnum<T>(names.Item1, names.Item2, @default, findByOrganization);
        }

        public T GetEnum<T>(string type, string lic, T @default = default(T), bool findByOrganization = false) where T : struct
        {
            var value = GetValue(type, lic, findByOrganization);

            if (value == null)
                return @default;

            return Enums<T>.Parse(value);
        }

        public bool GetBoolean(string name, bool @default = false)
        {
            var names = Split(name);
            return GetBoolean(names.Item1, names.Item2, @default);
        }

        public bool GetBoolean(string type, string lic, bool @default = false)
        {
            var value = GetListOfValue(type, lic);

            if (string.IsNullOrEmpty(value?.DisplayValue))
                return @default;

            var intValue = ObjectUtils.ToNullableInteger(value?.DisplayValue);

            if (intValue != null)
                return Convert.ToBoolean(intValue);

            return Convert.ToBoolean(value?.DisplayValue);
        }

        public decimal GetDecimal(string name, decimal @default = 0)
        {
            var names = Split(name);
            var value = GetStringValue(names.Item1, names.Item2);

            return value != null ? decimal.Parse(value) : @default;
        }

        public decimal? GetDecimalOrNull(string name)
        {
            var names = Split(name);
            var value = GetStringValue(names.Item1, names.Item2);

            return value != null ? decimal.Parse(value) : default(Decimal);
        }

        private Tuple<string, string> Split(string name)
        {
            var names = name.Split(new[] { "." }, 2, StringSplitOptions.RemoveEmptyEntries);
            if (names.Count() == 2)
            {
                var type = names[0];
                var lic = names[1];

                return new Tuple<string, string>(type, lic);
            }
            else if (names.Count() == 1)
            {
                var lic = names[0];

                return new Tuple<string, string>(null, lic);
            }

            return new Tuple<string, string>(null, null);
        }

        private string GetStringValue(string type, string lic)
        {
            ListOfValue value = null/* TODO Change to default(_) if this is not a reference type */;

            if (type == null)
                value = _values?.FirstOrDefault(f => f.LIC == lic);
            else
                value = _values?.FirstOrDefault(f => f.LIC == lic & f.Type == type);

            return value?.DisplayValue;
        }

        private ListOfValue GetListOfValue(string type, string lic)
        {
            return _values?.FirstOrDefault(f => f.Type == type & f.LIC == lic);
        }
    }
}